### PR TITLE
feat: Explain React Router integration for Vaadin Flow

### PR DESCRIPTION
**Pull request by mshabarov**
Related-to https://github.com/vaadin/platform/issues/4710

### CODE REVIEW

File: articles/create-ui/templates/using-react.asciidoc [hunk 1]
```diff
@@ -0,0 +1,126 @@
+---
+title: Use React library
+description: How to create a hybrid Vaadin applications using React
+order: 200
+---
+
+[role="since:com.vaadin:vaadin@V24.3"]
+= Vaadin Hybrid Applications With React
```
Comment (tarekoraby) [inline articles/create-ui/templates/using-react.asciidoc:8]:
```suggestion
= Flow-Hilla Hybrid Applications
```

File: articles/create-ui/templates/using-react.asciidoc [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+Vaadin allows to develop a hybrid application that leverages Vaadin Flow, React, and the https://hilla.dev/docs/react[Hilla framework].
+This gives an opportunity to combine Vaadin Flow components and routes with the ones written in React in one single application.
```
Comment (tarekoraby) [inline articles/create-ui/templates/using-react.asciidoc:11]:
```suggestion
https://hilla.dev/docs/react[Hilla] is a framework by Vaadin for building reactive web applications on Java backends. It seamlessly integrates a [React](https://reactjs.org/) TypeScript frontend with a [Spring Boot](https://spring.io/projects/spring-boot) backend.

You can develop hybrid applications that leverage Vaadin Flow and Hilla. This allows you to combine Vaadin Flow components and routes with the ones written in React in one single application.
```

File: articles/create-ui/templates/using-react.asciidoc [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+This section demonstrates how to create a hybrid application from the https://github.com/vaadin/skeleton-starter-hilla-react[React+Hilla application].
```
Comment (tarekoraby) [inline articles/create-ui/templates/using-react.asciidoc:13]: Can we provide the opposite: instructions for adding Hilla to existing Flow apps? This article will be mostly read by existing Flow developers who might be interested in adding Hilla to their existing Flow apps.

**Review by tarekoraby — APPROVED** [review]

File: articles/create-ui/templates/using-react.asciidoc (hunk 3 above)
Comment (caalador) [reply articles/create-ui/templates/using-react.asciidoc:13]: This should probably have 2 routes to choose from. `Add Flow to Hilla` and `Add Hilla to Flow`

File: articles/create-ui/templates/using-react.asciidoc [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+If this configuration is missed and the feature flag is enabled, Vaadin throws a runtime exception in build time.
+
+.Route configuration and React dependencies
+[NOTE]
+Vaadin creates [filename]`frontend/App.tsx` and [filename]`frontend/routes.tsx` files automatically, if they are missing, as well as internal `Frontend/generated/flow/Flow.tsx` file and adds React front-end dependencies, such as `react`, `react-dom` and `react-router-dom`.
+
+== Add Vaadin Flow routes
+
+Add Vaadin Flow routes, as described in https://vaadin.com/docs/latest/routing/route[Routing & Navigation].
+
```
Comment (caalador) [inline articles/create-ui/templates/using-react.asciidoc:121]: Missing discussion-id
```[discussion-id]`Random UUID` ```

File: articles/create-ui/templates/using-react.asciidoc [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+import {serverSideRoutes} from "Frontend/generated/flow/Flow";
+
+// ...
+export const routes = [
+    // Hilla routes definitions
+    ...serverSideRoutes
+]
+----
+
+.Import server-side routes for CRM tutorial
```
Comment (caalador) [inline articles/create-ui/templates/using-react.asciidoc:89]: This should have some highlight that this is an **example** on the base CRM routes.tsx

File: articles/create-ui/templates/using-react.asciidoc [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        children: [
+            { path: '/', element: <ContactsView />, handle: { title: 'Contacts' } },
+            { path: '/about', element: <AboutView />, handle: { title: 'About' } },
+            ...serverSideRoutes
+        ],
+    },
+];
+----
+
+Adding `serverSideRoutes` is crucial to let React Router navigate to server-side (Vaadin Flow) routes from React routes.
```
Comment (caalador) [inline articles/create-ui/templates/using-react.asciidoc:111]:
```suggestion
Adding `serverSideRoutes` is required for React Router to be able to navigate to server-side (Vaadin Flow) routes.
```
A big point here is also that to navigate at all to server router using react router you need the serverSideRoutes

File: articles/create-ui/templates/using-react.asciidoc [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+If this configuration is missed and the feature flag is enabled, Vaadin throws a runtime exception in build time.
+
+.Route configuration and React dependencies
+[NOTE]
+Vaadin creates [filename]`frontend/App.tsx` and [filename]`frontend/routes.tsx` files automatically, if they are missing, as well as internal `Frontend/generated/flow/Flow.tsx` file and adds React front-end dependencies, such as `react`, `react-dom` and `react-router-dom`.
```
Comment (caalador) [inline articles/create-ui/templates/using-react.asciidoc:116]:
```suggestion
Vaadin creates [filename]`frontend/App.tsx` and [filename]`frontend/routes.tsx` files automatically, if they are missing, as well as the internal `Frontend/generated/flow/Flow.tsx` file. Also React dependencies, such as `react`, `react-dom` and `react-router-dom`, are added to the `package.json` file and installed.
```

File: articles/create-ui/templates/using-react.asciidoc [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+Enable it by adding the following parameter into [filename]`src/main/resources/vaadin-featureflags.properties`:
+
+.`vaadin-featureflags.properties`
+[source,properties]
+----
+com.vaadin.experimental.reactRouter=true
+----
+
+== Add the server-side routes entry
```
Comment (caalador) [inline articles/create-ui/templates/using-react.asciidoc:73]:
```suggestion
== Add Server-side Routes Target
```

File: articles/create-ui/templates/using-react.asciidoc [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+            <scope>import</scope>
+        </dependency>
+        <!-- Other pom imports -->
+    </dependencies>
+</dependencyManagement>
+----
+
+Add the `vaadin-core` / `vaadin` dependency:
+
+.Vaadin and Hilla dependencies
```
Comment (caalador) [inline articles/create-ui/templates/using-react.asciidoc:51]:
```suggestion
.Vaadin dependency
```
As this sample has no hilla dependencies.

File: articles/create-ui/templates/using-react.asciidoc [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+Vaadin allows to develop a hybrid application that leverages Vaadin Flow, React, and the https://hilla.dev/docs/react[Hilla framework].
+This gives an opportunity to combine Vaadin Flow components and routes with the ones written in React in one single application.
+
+This section demonstrates how to create a hybrid application from the https://github.com/vaadin/skeleton-starter-hilla-react[React+Hilla application].
+
+== Add Vaadin And Hilla Dependencies
+
+Open your `pom.xml` file and add the necessary dependencies for Vaadin Flow:
+
+.Vaadin and Hilla versions
```
Comment (caalador) [inline articles/create-ui/templates/using-react.asciidoc:19]:
```suggestion
.Added Vaadin bom
```
As the hilla version and bom should already exist as we are adding to a hilla project

File: articles/create-ui/templates/using-react.asciidoc (hunk 3 above)
Comment (tarekoraby) [reply articles/create-ui/templates/using-react.asciidoc:13]: We can have both. But `Add Hilla to Flow` is the more relevant path for the Flow docs IMO.

File: articles/create-ui/templates/using-react.asciidoc
[the remark this answers is not in the document]
Comment (mshabarov) [reply articles/create-ui/templates/using-react.asciidoc:1]: For me, as Java and Vaadin developer, using a React is one more way of creating front-end views, i.e. fits into templates topic.
Adding it to "Start a Project" as one more tab looks weird, as Maven/Gradle cannot be grouped with Hilla and Hybrid apps, it's a different thing.
"Integrations" might be a good trade-off for now, when we have a separate docs for Hilla.

[the remark this answers is not in the document]
Comment (mshabarov) [reply articles/create-ui/templates/using-react.asciidoc:1]: Moved to "Integrations" with title "Hilla"

File: articles/create-ui/templates/using-react.asciidoc (hunk 4 above)
Comment (mshabarov) [reply articles/create-ui/templates/using-react.asciidoc]: added

File: articles/create-ui/templates/using-react.asciidoc (hunk 8 above)
Comment (mshabarov) [reply articles/create-ui/templates/using-react.asciidoc]: done

File: articles/create-ui/templates/using-react.asciidoc (hunk 7 above)
Comment (mshabarov) [reply articles/create-ui/templates/using-react.asciidoc]: done

File: articles/create-ui/templates/using-react.asciidoc (hunk 5 above)
Comment (mshabarov) [reply articles/create-ui/templates/using-react.asciidoc]: Reworded and added a link to CRM example.

**Review by caalador — APPROVED** [review]

**Review by mshabarov — COMMENTED** [review]
Last additions for "Adding Hilla to Flow app" looks good to me and overall article is good enough IMO.

**Comment by mshabarov** [comment]
@russelljtdyer @jouni Could you please help us with the language checking? We'd like to have this article live on Vaadin website until we publish a blogpost about Vaadin 24.3 (planned for tomorrow).

**Pull request merged**